pkg/gluetikv/glue.go

```diff
@@ -57,15 +60,28 @@ func (Glue) Record(name string, val uint64) {
 }
 
 type progress struct {
-	ch chan<- struct{}
+	ch     chan<- struct{}
+	closed int32
 }
 
 // Inc implements glue.Progress.
 func (p progress) Inc() {
+	if atomic.LoadInt32(&p.closed) != 0 {
+		log.Warn("proposing a closed progress")
+		return
+	}
+	// there might be buggy if the thread is yielded here.
+	// however, there should not be gosched, at most time.
+	// so send here probably is safe, even not totally safe.
+	// since adding an extra lock should be costly, we just be optimistic.
+	// (Maybe a spin lock here would be better?)
 	p.ch <- struct{}{}
```

Comment:
it is possible to just make this entire progress structure an atomic integer instead of a channel, like https://github.com/pingcap/tidb/blob/c7651f02be36b5faf95a204d20f8859eb9c2080a/executor/brie.go#L49-L73? tbh i don't see any reason why we use a channel in the first place, since we're using a 1-second timer to update the progress bar in any case.

Reply:
well there would be more change than this pr itself if we refactor the progess struct here. how about leaving a TODO here and leave it another PR?

Reply:
(But I could have a try this PR, anyway)